File: 307-range-sum-query-mutable/307-range-sum-query-mutable.cpp
class NumArray {
public:
    vector<int>seg;
    int n;
    NumArray(vector<int>& nums) {
      if (!nums.size()) return ;
        seg.assign(nums.size() * 4, 0);
        n = nums.size();
        build(nums, 0, 0, n - 1);   
    }
    void build(vector<int>& nums,int i , int l , int r)
    {
        if(l==r)
        {
            seg[i] = nums[l];
        }
        else{
        int mid = l + (r - l) / 2;
        build(nums,2*i+1,l,mid);
        build(nums, 2*i+2,mid+1,r);
            
        seg[i] = seg[2*i+1] + seg[2*i+2];
        }
    }
    
    void update(int i,int l,int r,int index,int val)
    {
        if(l==r)
        {
            seg[i] = val;
        }
        else
        {
           int mid = l + (r - l) / 2;
            if(index<=mid)
            {
                update(2*i+1,l,mid,index,val);
            }
            else
            {
                update(2*i+2,mid+1,r,index,val);
            }
            seg[i] = seg[2*i+1] + seg[2*i+2];
            
        }
    }
    void update(int index, int val) {
       update(0,0,n-1,index,val); 
    }
    
    int sumRange(int i, int l, int r, int left, int right)
    {
        if(r < left or l > right)
        {
            return 0;
        }
        if(l >=left && right >= r)
        {
            return seg[i];
        }
        int mid = l + (r - l) / 2;
        return sumRange(2*i+1,l,mid,left,right) + sumRange(2*i+2,mid+1,r,left,right);
    }
    
    int sumRange(int left, int right) {
     return  sumRange(0,0,n-1,left,right); 
    }

};

File: 307-range-sum-query-mutable/307-range-sum-query-mutable.java
// Fenwick Tree
class NumArray {
    
    class ftree{
        int []ft;
        ftree(int []nums){
            ft= new int[nums.length+1];
            for(int i=0;i<or.length;i++){
                int pos = i+1;
                update(pos,nums[i]);
            }
        }
        private int rsb(int x){
            return x&-x;
        }
        int getsum (int pos){
            int val=0;
            while(pos>0){
                val+=ft[pos];
                pos-=rsb(pos);
            }
            return val;
        }
        void update(int pos, int delta){
            while(pos<ft.length){
                ft[pos]+=delta;
                pos = pos+rsb(pos);
            }
            
        }
        
    }
    int [] or;
    ftree node;
    public NumArray(int[] nums) {
        or= nums;
        node = new ftree(nums);
        
    }
    
    public void update(int index, int val) {
        int delta = val-or[index];
        node.update(index+1,delta);
        or[index]=val;
    }
    
    public int sumRange(int left, int right) {
        // now we have to calculate the sum
        int lpos = left+1;
        int rpos = right+1;
        int leftSum = node.getsum(lpos-1);
        int rightSum = node.getsum(rpos);
        return rightSum-leftSum;
    }
}

/**
 * Your NumArray object will be instantiated and called as such:
 * NumArray obj = new NumArray(nums);
 * obj.update(index,val);
 * int param_2 = obj.sumRange(left,right);
 */